src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp

```diff
@@ -16,112 +16,144 @@
  * with this program. If not, see <http://www.gnu.org/licenses/>.
  */
 
-/* ScriptData
-SDName: Boss_Baroness_Anastari
-SD%Complete: 90
-SDComment: MC disabled
-SDCategory: Stratholme
-EndScriptData */
-
-#include "ScriptMgr.h"
 #include "InstanceScript.h"
+#include "ObjectAccessor.h"
+#include "Player.h"
+#include "ScriptMgr.h"
 #include "ScriptedCreature.h"
 #include "stratholme.h"
 
 enum Spells
 {
-    SPELL_BANSHEEWAIL       = 16565,
-    SPELL_BANSHEECURSE      = 16867,
-    SPELL_SILENCE           = 18327
-    //SPELL_POSSESS           = 17244
+    SPELL_BANSHEEWAIL           = 16565,
+    SPELL_BANSHEECURSE          = 16867,
+    SPELL_SILENCE               = 18327,
+    SPELL_POSSESS               = 17244,    // The charm
+    SPELL_POSSESSED             = 17246     // The damage buff
 };
 
-class boss_baroness_anastari : public CreatureScript
+enum BaronessAnastariEvents
 {
-public:
-    boss_baroness_anastari() : CreatureScript("boss_baroness_anastari") { }
-
-    CreatureAI* GetAI(Creature* creature) const override
-    {
-        return GetStratholmeAI<boss_baroness_anastariAI>(creature);
-    }
-
-    struct boss_baroness_anastariAI : public ScriptedAI
-    {
-        boss_baroness_anastariAI(Creature* creature) : ScriptedAI(creature)
-        {
-            Initialize();
-            instance = me->GetInstanceScript();
-        }
+    EVENT_SPELL_BANSHEEWAIL     = 1,
+    EVENT_SPELL_BANSHEECURSE    = 2,
+    EVENT_SPELL_SILENCE         = 3,
+    EVENT_SPELL_POSSESS         = 4,
+    EVENT_INVISIBLE             = 5,
+    EVENT_CHECK_POSSESSED       = 6
+};
 
-        void Initialize()
-        {
-            BansheeWail_Timer = 1000;
-            BansheeCurse_Timer = 11000;
-            Silence_Timer = 13000;
-            //Possess_Timer = 35000;
-        }
 
-        InstanceScript* instance;
+struct boss_baroness_anastari : public BossAI
+{
+    boss_baroness_anastari(Creature* creature) : BossAI(creature, TYPE_BARONESS) { }
 
-        uint32 BansheeWail_Timer;
-        uint32 BansheeCurse_Timer;
-        uint32 Silence_Timer;
-        //uint32 Possess_Timer;
+    EventMap _events;
+    ObjectGuid _possessedTargetGuid;
 
-        void Reset() override
+    void Reset() override
+    {
+        if (!me->IsVisible())
         {
-            Initialize();
+            if (Player* possessedTarget = ObjectAccessor::FindConnectedPlayer(_possessedTargetGuid))
+            {
+                if (possessedTarget->IsAlive())
+                {
+                    possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESSED);
+                    possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESS);
+                    me->SetReactState(REACT_AGGRESSIVE);
+                    me->SetVisible(true);
```

Comment:
If the possessed target is dead for some reason, you want to stay invisible? Sounds wrong.

Reply:
That's why I'm checking if target is alive
```
                    if (possessTarget->IsAlive())
                    {
                        me->CastStop();
                        DoCast(possessTarget, SPELL_POSSESS);
                        DoCast(possessTarget, SPELL_POSSESSED);
                        me->SetReactState(REACT_PASSIVE);
                        me->SetVisible(false);
                        _possessedTargetGuid = possessTarget->GetGUID();
                        _events.ScheduleEvent(EVENT_CHECK_POSSESSED, 0s);
                    }
```

Reply:
What if it dies between then and now?